keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1004 @@
+consumption by workloads, whether directly, via Kubelet, or otherwise.
+
+Every signer corresponds to zero, one, or more TrustAnchorSets objects.  The
+name of the TrustAnchorSets objects are arbitrary; the association is managed
+entirely via the `spec.signerName` field, which is immutable.
+
+Since a given signer could have multiple TrustAnchorSets objects, clients
+looking for a particular named set X must take the union of all `spec.sets`
+entries where `spec.sets[i].name == X`.
+
+#### API Object Definition
+
+```go
+// TrustAnchorSets packages up trust anchors.
+//
+// Each TrustAnchorSets object is owned by a particular signer, which
+// updates the trust anchors as the roots of trust for that signer change.
+//
+// A TrustAnchorSets object wraps up multiple trust anchor sets, each of
+// which in turn is a set of one or more trust anchors (individual
+// certificates).  Most use cases will only require a single trust
+// anchor set, conventionally named `default`.  Multiple uniquely-named trust
+// anchor sets will likely only be required in mutual TLS use cases with
+// federation between multiple meshes.
+type TrustAnchorSets struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec TrustAnchorSetsSpec `json:"spec"`
+  Status TrustAnchorSetsStatus `json:"status"`
+}
+
+type TrustAnchorSetsSpec struct {
+  // The name of the signer that owns this TrustAnchorSets object.
+  //
+  // Each signer may maintain multiple TrustAnchorSets objects.
+  SignerName string
+}
+
+type TrustAnchorSetsStatus struct {
+  // The list of trust anchor sets.
+  Sets []TrustAnchorSet
+
+  // The current conditions of the TrustAnchorSets object.
+  //
+  // There is one canonical condition, "Ready", that the signer uses
+  // to indicate to kubelet that the TrustAnchorSets object is up to date.
+  // This is used to keep pods from starting up with empty trust anchor sets
+  // while the signer is initializing.
+  Conditions []metav1.Condition
+}
+
+type TrustAnchorSet struct {
+  // The unique name of this trust anchor set.
+  Name string
+
+  // The individual trust anchors for this set.  Each entry is a single
+  // PEM-wrapped, DER-formatted X.509 certificate.
+  PEMTrustAnchors []string
+}
+```
+
+#### RBAC Interaction
+
+Like CertificateSigningRequests, it will be possible to grant RBAC permissions
+to TrustAnchorSets based on the associated signer name, rather than the name of
+the TrustAnchorSetsObject.
+
+For example, a controller that manages TrustAnchorSets would typically be
+allowed to do anything to TrustAnchorSets objects with the correct signerName:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: example-com-trustmanager
+rules:
+- apiGroups:
+  - certificates.k8s.io
+  resources:
+  - signers/trustanchorsets
+  resourceNames:
+  - example.com/my-signer-name # Or example.com/* for all signers in the example.com domain.
```

Comment:
This would not work for reads?

Reply:
Yes.  This has been reworked to use a fixed mapping between a signer name and a single TrustAnchorSets object.